VacuumChest

ferriarnus opened the pull request:
# Description

The Vacuum Chest. I'm gonna make this a draft so people can easily see it. Should I have a look at filters or not, since this can somewhat use them.

# Todo <!-- Remove this section if you're submitting an already-complete PR -->

# Vacuum Chest
- [x] Work with hopper? (general storage input/output)
- [ ] Model
- [ ] Filter?
- [ ] Gui layout

# Vacuum Chest
- [x] Work with tank? (general storage input/output)
- [ ] Model
- [x] xp liquid?
- [ ] Gui layout

# Checklist:

- [x] My code follows the style guidelines of this project (.editorconfig, most IDEs will use this for you)
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

<!-- Thanks to: https://embeddedartistry.com/blog/2017/08/04/a-github-pull-request-template-for-your-projects/ for the building blocks of this template -->

ferriarnus committed:
VacuumChest start

ferriarnus committed:
VacuumChest start

ferriarnus committed:
VacuumChest start

ferriarnus committed:
VacuumChest start

ferriarnus committed:
start xpvacuum

ferriarnus committed:
start xpvacuum

ferriarnus committed:
fixed fluidstackwidget

ferriarnus committed:
fixed fluidstackwidget

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/client/gui/screen/VacuumChestScreen.java:
```diff
@@ -0,0 +1,58 @@
+	
+	@Override
+	protected void renderLabels(PoseStack pPoseStack, int pMouseX, int pMouseY) {
+		this.font.draw(pPoseStack, new TranslatableComponent("enderio.gui.filter"), 8, this.imageHeight - 134, 4210752);
```
Use EIOLang as this key will also be used by conduits

justliliandev replied:
Use EIOLang as this key will also be used by conduits

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/client/gui/screen/XPVacuumScreen.java:
```diff
@@ -0,0 +1,53 @@
+	@Override
+	public void render(PoseStack pPoseStack, int pMouseX, int pMouseY, float pPartialTicks) {
+		super.render(pPoseStack, pMouseX, pMouseY, pPartialTicks);
+		font.draw(pPoseStack, this.getMenu().getBlockEntity().getRange() +"", leftPos + imageWidth - 8 - 12 - 20 - 2 - 8, topPos + 50, 0);
```
```suggestion
		font.draw(pPoseStack, this.getMenu().getBlockEntity().getRange() + "", leftPos + imageWidth - 8 - 12 - 20 - 2 - 8, topPos + 50, 0);
```
add space

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/client/gui/widget/FluidStackStaticWidget.java:
```diff
@@ -0,0 +1,75 @@
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.fluids.capability.templates.FluidTank;
+
+public class FluidStackStaticWidget extends EIOWidget {
```
Move FluidRenderCode to either some common class or a Helper class, to reduce duplicated code

ferriarnus replied:
I get why you want this, but I really don't see how to do it properly.

justliliandev replied:
Alrighty.
I can take a look at it, once it's merged

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/client/gui/widget/FluidStackStaticWidget.java:
```diff
@@ -0,0 +1,75 @@
+
+                    int atlasWidth = (int)(sprite.getWidth() / (sprite.getU1() - sprite.getU0()));
+                    int atlasHeight = (int)(sprite.getHeight() / (sprite.getV1() - sprite.getV0()));
+                    //blit(poseStack, x, y, displayOn.getBlitOffset(), sprite.getU0()*atlasWidth, sprite.getV0()*atlasHeight, sprite.getWidth(), sprite.getHeight(), atlasWidth, atlasHeight);
```
why is this line commented out

ferriarnus replied:
Probably me messing with different blit metods and forgot to fully remove it

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/client/gui/widget/FluidStackStaticWidget.java:
```diff
@@ -0,0 +1,75 @@
+
+                    int atlasWidth = (int)(sprite.getWidth() / (sprite.getU1() - sprite.getU0()));
+                    int atlasHeight = (int)(sprite.getHeight() / (sprite.getV1() - sprite.getV0()));
+                    //blit(poseStack, x, y, displayOn.getBlitOffset(), sprite.getU0()*atlasWidth, sprite.getV0()*atlasHeight, sprite.getWidth(), sprite.getHeight(), atlasWidth, atlasHeight);
```
why is this line commented out?

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/block/EnhancedMachineBlock.java:
```diff
@@ -129,7 +131,7 @@ public BlockEntity newBlockEntity(BlockPos pPos, BlockState pState) {
         return null;
     }
-
+    
```
undo change

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/VacuumChestBlockEntity.java:
```diff
@@ -0,0 +1,132 @@
+	
+	@Override
+	public void tick() {
+		super.tick();
+		if (this.getRedstoneControl().isActive(level.hasNeighborSignal(worldPosition))) {
```
switch order to allow for instant push out

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/VacuumChestBlockEntity.java:
```diff
@@ -0,0 +1,132 @@
+	
+	private void collectItems(Level level, BlockPos pos, int range) {
+		AABB area = new AABB(pos).inflate(range);
+		List<ItemEntity> itemEntities = level.getEntitiesOfClass(ItemEntity.class, area, e -> true); //TODO filter logic
```
cache that result for a few ticks (maybe 5(?)) as this is a rather expensive operation. Also don't module just the global ticktime, but rather modulo the position hash or be hash, so that multiple vacuum chests don't search for item entities in the same tick, balancing that load over multiple ticks

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/VacuumChestBlockEntity.java:
```diff
@@ -0,0 +1,132 @@
+		}
+	}
+	
+	private void collectItems(Level level, BlockPos pos, int range) {
```
Move that code to a common place for combined usage with magnet

ferriarnus replied:
Where would I put it. The code is also not fully the same, cause it needs to act differently for each of them. I'm unsure how I would properly do it. Maybe split it in 2-3 methods?

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/VacuumChestBlockEntity.java:
```diff
@@ -0,0 +1,132 @@
+	public void decreasseRange() {
+		if (this.range > 0) {
+			this.range--;
+			System.out.println(this.range);
```
remove sout

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/VacuumChestBlockEntity.java:
```diff
@@ -0,0 +1,132 @@
+	public void increasseRange() {
+		if (this.range < MAX_RANGE) {
+			this.range++;
+			System.out.println(this.range);
```
remove sout

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/XPVacuumBlockEntity.java:
```diff
@@ -0,0 +1,128 @@
+import net.minecraftforge.fluids.capability.IFluidHandler.FluidAction;
+
+public class XPVacuumBlockEntity extends MachineBlockEntity {
+	private static final double COLLISION_DISTANCE_SQ = 1 * 1;
```
remove one tab

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/XPVacuumBlockEntity.java:
```diff
@@ -0,0 +1,128 @@
+	private static final double COLLISION_DISTANCE_SQ = 1 * 1;
+    private static final double SPEED = 0.025;
+    private static final double SPEED_4 = SPEED ;
+	private static final int MAX_RANGE = 6;
```
here too

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/XPVacuumBlockEntity.java:
```diff
@@ -0,0 +1,128 @@
+			BlockState pBlockState) {
+		super(MachineTier.STANDARD, pType, pWorldPosition, pBlockState);
+		this.fluidTank =  new FluidTankMaster(Integer.MAX_VALUE, getIoConfig()); //that seems quite large?
+		addDataSlot(new FluidStackDataSlot(() -> fluidTank.getFluidInTank(0), fluidTank::setFluid, SyncMode.WORLD));
```
as only one fluid should be here, maybe think about using an in integer dataslot as the fluid isn't required to be synced

ferriarnus replied:
Good idea, hadn't considered that.

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/XPVacuumBlockEntity.java:
```diff
@@ -0,0 +1,128 @@
+
+	public XPVacuumBlockEntity(BlockEntityType<?> pType, BlockPos pWorldPosition,
+			BlockState pBlockState) {
+		super(MachineTier.STANDARD, pType, pWorldPosition, pBlockState);
```
fix spacing

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/XPVacuumBlockEntity.java:
```diff
@@ -0,0 +1,128 @@
+	}
+	
+	public FluidTankMaster getFluidTank() {
+		return fluidTank;
```
fix spacing

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/XPVacuumBlockEntity.java:
```diff
@@ -0,0 +1,128 @@
+        fluidTank.readFromNBT(pTag.getCompound("Fluids"));
+    }
+
+	@Override
```
fix spacing

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/XPVacuumBlockEntity.java:
```diff
@@ -0,0 +1,128 @@
+	
+	@Override
+	public void tick() {
+		super.tick();
+		if (this.getRedstoneControl().isActive(level.hasNeighborSignal(worldPosition))) {
```
switch order for instant push out

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/XPVacuumBlockEntity.java:
```diff
@@ -0,0 +1,128 @@
+	private void collectXP(Level level, BlockPos pos, int range) {
+		AABB area = new AABB(pos).inflate(range);
+		List<ExperienceOrb> xpEntities = level.getEntitiesOfClass(ExperienceOrb.class, area);
+		for (ExperienceOrb xpe: xpEntities) { //magnet code
```
move to common helper

ferriarnus committed:
some fixes

ferriarnus committed:
some fixes

ferriarnus committed:
fix spacing

ferriarnus committed:
fix spacing

ferriarnus committed:
Update VacuumChestBlockEntity.java

ferriarnus committed:
Update VacuumChestBlockEntity.java

ferriarnus committed:
tick fix

Making not all blocks tick at the same time

ferriarnus committed:
tick fix

Making not all blocks tick at the same time

justliliandev commented on enderio-base/src/main/java/com/enderio/base/common/util/AttractionUtil.java:
```diff
@@ -0,0 +1,62 @@
+     * @param collisionDistanceSq
+     * @return if the entity is inside the collision distance.
+     */
+    public static boolean hasReachedPos(Entity entity , BlockPos pos, double speed, double speed4, double collisionDistanceSq) {
```
rename, because the name implies, that it just checks something, but doesn't do something.
moveToPos would be a better name imo.
Also the Description of the method describes it's return value, so remove the duplicate in the description

justliliandev commented on enderio-base/src/main/java/com/enderio/base/common/util/AttractionUtil.java:
```diff
@@ -0,0 +1,62 @@
+     * @return if the entity is inside the collision distance.
+     */
+    public static boolean hasReachedPos(Entity entity , double x, double y, double z, double speed, double speed4, double collisionDistanceSq) {
+        if (entity == null) { //If the entity no longer exists, return false
+            return false;
+        }
```
Entities are never null in this context

ferriarnus replied:
Hmm, true. My intention was to ignore entities that might not existed anymore. Would another check "isremoved" or something like that work?

justliliandev replied:
Yes, but maybe cache them with a weak reference and check if it still exists and do an isRemoved check

ferriarnus replied:
I get why you say that, but I'm not certain it is worth it in this case. The list is made every 5 ticks, and if I want to make a list of weak reference I need to iterate over the already existing list again. I'm not certain if we gain anything by using a weak reference.

justliliandev commented on enderio-base/src/main/java/com/enderio/base/common/util/AttractionUtil.java:
```diff
@@ -0,0 +1,62 @@
+            double deltaZ = mov.z + z * adjustedSpeed;
+            double deltaY;
+            if (y > 0) {
+              //if items are below, raise them to player level at a fixed rate
```
Rename things in comment:
items -> entity
player -> target

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/VacuumChestBlockEntity.java:
```diff
@@ -0,0 +1,115 @@
+    }
+    
+    private void collectItems(Level level, BlockPos pos, int range) {
+        if ((level.getGameTime() + pos.asLong()) % 5 == 0 || this.itemEntities.isEmpty()) {
```
remove the isEmpty check as it will be done each tick, if no itemEntities are Available. Instead do this in BlockEntitiy#onLoad

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/XPVacuumBlockEntity.java:
```diff
@@ -0,0 +1,112 @@
+        this.range = range;
+    }
+    
+    public void decreasseRange() {
```
fix spelling error: decrease and increase

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/menu/XPVacuumMenu.java:
```diff
@@ -0,0 +1,32 @@
+    public XPVacuumMenu(XPVacuumBlockEntity blockEntity, Inventory inventory, int pContainerId) {
+        super(blockEntity, inventory, MachineMenus.XP_VACUUM.get(), pContainerId);
+        if (blockEntity != null) {
+            
```
Remove this check or explain why it's empty

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/VacuumChestBlockEntity.java:
```diff
@@ -0,0 +1,115 @@
+        this.range = range;
+    }
+    
+    public void decreasseRange() {
```
fix spelling

ferriarnus committed:
more fixes and spelling

ferriarnus committed:
more fixes and spelling

ferriarnus committed:
added WeakReference

ferriarnus committed:
added WeakReference

justliliandev commented on enderio-base/src/main/java/com/enderio/base/common/item/tool/ElectromagnetItem.java:
```diff
@@ -1,18 +1,20 @@
 
+import java.util.List;
+
+import javax.annotation.Nonnull;
```
remove nonnull import, as we usually mark nullable things with nullable instead of non null things with nonnnull

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/VacuumChestBlockEntity.java:
```diff
@@ -0,0 +1,133 @@
+        if ((level.getGameTime() + pos.asLong()) % 5 == 0) {
+            AABB area = new AABB(pos).inflate(range);
+            for (ItemEntity ie:level.getEntitiesOfClass(ItemEntity.class, area, e -> true)) {
+                this.itemEntities.add(new WeakReference<ItemEntity>(ie));
```
```suggestion
                this.itemEntities.add(new WeakReference(ie));
```
this generic isn't necessary, also move the todo comment  to e-> true

ferriarnus replied:
It may not be needed, but my ide does warn me about it. It shouldn't matter though?

justliliandev commented on enderio-base/src/main/java/com/enderio/base/common/util/AttractionUtil.java:
```diff
@@ -0,0 +1,48 @@
+     * Pulls the entity closer to the given x, y and z coordinates. Once it is in inside the collision distance, the function will return true, otherwise it returns false.
+     */
+    public static boolean moveToPos(Entity entity, double x, double y, double z, double speed, double speed4, double collisionDistanceSq) {
+        if (entity.isRemoved()) { //If the entity no longer exists, return false
+            return false;
+        }
```
add this check to it's callers to remove the entity from the caching lists, if possible

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/VacuumChestBlockEntity.java:
```diff
@@ -0,0 +1,133 @@
+            for (ItemEntity ie:level.getEntitiesOfClass(ItemEntity.class, area, e -> true)) {
+                this.itemEntities.add(new WeakReference<ItemEntity>(ie));
+            }; //TODO filter logic
+        }
```
move this to a helper, because it duplicates code with collectItems

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/VacuumChestBlockEntity.java:
```diff
@@ -0,0 +1,133 @@
+    
+    private void collectItems(Level level, BlockPos pos, int range) {
+        if ((level.getGameTime() + pos.asLong()) % 5 == 0) {
+            AABB area = new AABB(pos).inflate(range);
```
clear list instead of just adding onto it

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/XPVacuumBlockEntity.java:
```diff
@@ -0,0 +1,130 @@
+    }
+    
+    private void collectXP(Level level, BlockPos pos, int range) {
+        if ((level.getGameTime() + pos.asLong()) % 5 == 0) {
+            AABB area = new AABB(pos).inflate(range);
+            for (ExperienceOrb ex:level.getEntitiesOfClass(ExperienceOrb.class, area)) {
+                this.xpEntities.add(new WeakReference<ExperienceOrb>(ex));
+            };
+        }
```
everything from VacuumChestBlockEntity
create common and generic helper for this and the itemEntity finding
parameters you probably will use: list, class, level, pos, range, filter (Predicate)

ferriarnus committed:
Merge branch 'vacuumchest' of https://github.com/ferriarnus/EnderIO-Rewrite118 into vacuumchest

ferriarnus committed:
more changes

made abstract vacuummachine class for the vacuum chest and xp vacuum

ferriarnus committed:
more changes

made abstract vacuummachine class for the vacuum chest and xp vacuum

Rover656 committed:
Some cleaning and move stuff to an API package

Rover656 committed:
Cleaned more and added a creative power block, doesn't push power yet.

Rover656 committed:
More recipe code cleaning

Rover656 committed:
Power pushing

Rover656 committed:
Added temporary graphic for the block

Rover656 committed:
Merge branch 'dev/1.18.x' into feature/powergen

Rover656 committed:
Moved the api package into enderio-base for now, it was causing IDE problems

Rover656 committed:
Mitigate a client freeze when removing a capacitor from a charged block

Rover656 committed:
Refactor the MachineBlockEntity class

Rover656 committed:
Minor cleaning

Rover656 committed:
Draft of new capacitor system.

Rover656 committed:
Unused import, but the game ran successfully!

Rover656 committed:
Changed to a functional interface for scaling.

Scalers themselves will need implementing and testing for balance purposes in the future.

Rover656 committed:
Added the original energy scaler. Next up is polishing machine integration.

Rover656 committed:
Move the capacitor keys into constructor instead, its cleaner.

Rover656 committed:
Add a nicity method

Rover656 committed:
Small cleaning

Rover656 committed:
Add "none" capacitor

Rover656 committed:
Tidy

Rover656 committed:
Merge pull request #2 from Rover656/concept/new-caps

New Capacitor System

Rover656 committed:
Starting stirling generator and fixed a simple machine bug.

Rover656 committed:
Stirling generator implementation

Rover656 committed:
Stirling generator model. Omitting simple for now. See desc.

Texture sizes in 1.12 were 32x32, we're striving to hit 16x16 in 1.18. Textures for these machines are going to need an art pass before making the final models.

Rover656 committed:
Fix readme

ferriarnus committed:
fix merge conflicts

ferriarnus committed:
Merge branch 'vacuumchest' of https://github.com/ferriarnus/EnderIO-Rewrite118 into vacuumchest

ferriarnus committed:
Merge branch 'dev/1.18.x' into vacuumchest

ferriarnus committed:
Merge branch 'dev/1.18.x' into vacuumchest

ferriarnus committed:
first inventory changes

ferriarnus committed:
Merge branch 'dev/1.18.x' into vacuumchest

ferriarnus committed:
Update XPVacuumBlockEntity.java

Rover656 commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/VacuumChestBlockEntity.java:
```diff
@@ -0,0 +1,89 @@
+	
+	// Slot config
+	
+	public Builder extractableGUISlot(Builder builder, int count) {
```
I might move this into the builder, but dont worry about doing that in this PR.

Rover656 commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/VacuumChestBlockEntity.java:
```diff
@@ -0,0 +1,89 @@
+            
+            @Override
+            public ItemStack insertItem(int slot, ItemStack stack, boolean simulate) {
+                if (slot == 27) { //TODO filter slot type
```
Could you make this use a predicate? The slots support a BiPredicate which gives you the item stack and the slot number, could do `(stack, slot) -> slot != 27`

Rover656 commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/XPVacuumBlockEntity.java:
```diff
@@ -0,0 +1,113 @@
+    
+    public XPVacuumBlockEntity(BlockEntityType<?> pType, BlockPos pWorldPosition, BlockState pBlockState) {
+        super(pType, pWorldPosition, pBlockState, ExperienceOrb.class);
+     // Create fluid tank.
```
indentation

ferriarnus committed:
Merge branch 'dev/1.18.x' into vacuumchest

Rover656 commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/base/VacuumMachineEntity.java:
```diff
@@ -0,0 +1,125 @@
+		add2WayDataSlot(new IntegerDataSlot(() -> this.getRange(), this::setRange, SyncMode.GUI));
+	}
+	
+	@Override
```
Note for myself: Might need to add a commonTick() to make this easier, or move this code into another method to reduce duplication

Rover656 commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/base/VacuumMachineEntity.java:
```diff
@@ -0,0 +1,125 @@
+        }
+        super.onLoad();
+    }
+	
```
newlines

Rover656 commented on enderio-machines/src/main/java/com/enderio/machines/common/energy/EnergyTransferMode.java:
```diff
@@ -0,0 +1,7 @@
+package com.enderio.machines.common.energy;
+
+public enum EnergyTransferMode {
```
What is this?

ferriarnus replied:
I did not make this? Might be some git weirdness? It blames you ;p

Rover656 replied:
Ahh might be part of the old energy rework, go ahead and see if its used anywhere, if not just delete the file :P (although try merging upstream first to see if it goes away)

Rover656 commented on enderio-machines/src/main/java/com/enderio/machines/common/init/MachineMenus.java:
```diff
@@ -1,8 +1,22 @@
 package com.enderio.machines.common.init;
 
 import com.enderio.machines.EIOMachines;
-import com.enderio.machines.client.gui.screen.*;
```
I think wildcards are fine for this purpose honestly.

ferriarnus replied:
That's my ide, it does that on save. Is it really an issue?

Rover656 replied:
Nah not an issue, was more of a comment if it was done intentionally :)

Rover656 reviewed (changes requested):
Some minor comments, a lot is for my 1.19 review though.

ferriarnus committed:
fixes

filter slot + small fixes

Rover656 commented:
Closing in favour of #80
